pkg/storagev2/inmemstore/inmembatch.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package inmemstore
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/storagev2"
+)
+
+type batchOp struct {
+	delete bool
+	item   storage.Item
+}
+
+type Batch struct {
+	mu    sync.Mutex
+	ctx   context.Context
+	ops   map[string]batchOp
+	store *Store
+	done  bool
+}
+
+func (s *Store) Batch(ctx context.Context) (storage.Batch, error) {
+	return &Batch{
+		ctx:   ctx,
+		ops:   make(map[string]batchOp),
+		store: s,
+	}, nil
+}
+
+func (i *Batch) Put(item storage.Item) error {
+	if i.ctx.Err() != nil {
+		return i.ctx.Err()
+	}
+
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	key := getKeyString(item)
+
+	i.ops[key] = batchOp{
+		item: item,
+	}
+
+	return nil
+}
+
+func (i *Batch) Delete(key storage.Key) error {
+	if i.ctx.Err() != nil {
+		return i.ctx.Err()
+	}
+
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	sKey := getKeyString(key)
+
+	i.ops[sKey] = batchOp{
+		delete: true,
+	}
+
+	return nil
+}
+
+func (i *Batch) Commit() error {
+	if i.ctx.Err() != nil {
+		return i.ctx.Err()
+	}
+
```

Comment:
So in order to be sure 100% we need to also have the lock on the batch mutex. During commit we cannot handle any Put or Delete operations.

Reply:
fixed